pkg/webhook/podprobemarker/validating/probe_create_update_handler.go

```diff
@@ -0,0 +1,235 @@
+func (h *PodProbeMarkerCreateUpdateHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &appsv1alpha1.PodProbeMarker{}
+	err := h.Decoder.Decode(req, obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	var old *appsv1alpha1.PodProbeMarker
+	//when Operation is update, decode older object
+	if req.AdmissionRequest.Operation == admissionv1.Update {
+		old = new(appsv1alpha1.PodProbeMarker)
+		if err := h.Decoder.Decode(
+			admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Object: req.AdmissionRequest.OldObject}},
+			old); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+	allErrs := h.validatingPodProbeMarkerFn(obj, old)
+	if len(allErrs) != 0 {
+		return admission.Errored(http.StatusBadRequest, allErrs.ToAggregate())
+	}
+	return admission.ValidationResponse(true, "")
+}
+
+func (h *PodProbeMarkerCreateUpdateHandler) validatingPodProbeMarkerFn(obj, old *appsv1alpha1.PodProbeMarker) field.ErrorList {
+	//validate ppm.Spec
+	allErrs := validatePodProbeMarkerSpec(obj, field.NewPath("spec"))
+	// when operation is update, validating whether old and new pps conflict
+	if old != nil {
+		if !reflect.DeepEqual(old.Spec.Selector, obj.Spec.Selector) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("selector"), obj.Spec.Selector, "selector must be immutable"))
+		}
+	}
+	return allErrs
+}
+
+func validatePodProbeMarkerSpec(obj *appsv1alpha1.PodProbeMarker, fldPath *field.Path) field.ErrorList {
+	spec := &obj.Spec
+	allErrs := field.ErrorList{}
+	// selector
+	if spec.Selector == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "empty Selector is not valid for PodProbeMarker."))
+	} else {
+		allErrs = append(allErrs, validateSelector(spec.Selector, fldPath.Child("selector"))...)
+	}
+	// containerProbe
+	if len(spec.Probes) == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("probes"), spec.Selector, "empty probes is not valid for PodProbeMarker."))
+		return allErrs
+	}
+	uniqueProbe := sets.NewString()
+	for _, probe := range spec.Probes {
+		if probe.Name == "" || probe.ContainerName == "" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("probes"), spec.Selector, "probe name or containerName can't be empty in PodProbeMarker."))
+			return allErrs
+		}
+		if k8sNativePodConditions.Has(probe.Name) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("probes"), spec.Selector, fmt.Sprintf("probe name can't be %s", probe.Name)))
+			return allErrs
+		}
+		if uniqueProbe.Has(probe.Name) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("probes"), spec.Selector, "probe name must be unique in PodProbeMarker."))
+			return allErrs
+		}
+		uniqueProbe.Insert(probe.Name)
+		allErrs = append(allErrs, validateProbe(&probe.Probe, fldPath.Child("probe"))...)
+		uniquePolicy := sets.NewString()
+		for _, policy := range probe.MarkerPolicy {
```

Comment:
shall we check each probe contains at least one markerPolicy?

Reply:
MarkerPolicy should not be required, the current design of kruise-game should not fill in the MarkerPolicy.